### ISSUE TITLE
Managing and persisting user preferences

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#415]
# Preferences and persistence
This is a brief sketch of how I've been thinking about the preferences / config system, how the various levels of preferences fit together, and some thoughts on how we might manage client settings.

## Config levels

There are a number of configuration 'tiers', which override one another in a set order:

1. **Defaults**: the base level settings that ship with xi; these are included in `assets/defaults.toml`. These are always loaded first, and provide sane general-case defaults.
2. **Platform specific**: these are platform-specific overrides that ship with xi, in (for instance) `assets/windows.toml`. These override the default settings on a particular platform.
3. **General user preferences** (optional): these are persistent user preferences, which override defaults. When a user modifies preferences, they are modifying these. 
1. **Syntax specific (system)**: these are syntax-specific settings which are bundled with xi. These override even user preferences; if a user's preferences declare a desire for tabs over spaces, the yaml syntax (`assets/yaml.toml`) preferences (which want spaces over tabs) will be preferred.
1. **Syntax specific (user)** (optional): The user can optionally set their own syntax-specific settings, which will override all previous settings, and persist between application launches.
1. **Session overrides (system)**: finally, on a per-buffer basis xi-core may override some settings, for instance if we have a preference for spaces over tabs but open a file that is indented with tabs, we might override the user preference for that given file.
1. **Session overrides (user)**: through the RPC protocol, the client may override a given setting for the current buffer. These overrides are not persistent; they would be used to manually set a line-ending or encoding on a per-document, per-session basis.

Of these, only the third and fifth are user-defined and persistent; they're the focus of this discussion.

## Getting and persisting user preferences
Two of these tiers (general and syntax-specific) require us to persist user preferences. There are a few possible approaches to this:
1. **File based, core managed**: the classic unix text editor approach: all preferences are saved in preference files; files are watched and preferences reloaded on change. This is the easiest approach, and has pedigree, but it can confuse and frustrate newer users, and doesn't really respect platform conventions.

2. **RPC based, client managed** The opposite pole: the client manages all user settings, sending them to core at launch and then keeping core up to date about changes. This would involve RPC calls to the effect of:

    ```javascript
    {
        "method": "set_preferences",
        "params": {
            "user": {
                "font_size" 14
            }
            "python": {
                "translate_tabs_to_spaces": true,
                "tab_size": 2
            }
        }
    }
    ```
This approach gives the client total control over how preferences are managed, but also requires each client to implement a settings system.

3. **File based, shared** A compromise; preferences are file based, but instead of editing those files directly, the user has some platform idiomatic preferences interface, and the client writes the user's preferences to disk, where changes are picked up by core. This approach is flexible (only clients that want to do this have to) but it's also fairly inelegant.

4. **File backed, RPC modified** Another compromise; core stores all preferences, and sends them to the client at launch. The client can present an idiomatic UI to the user, and preference changes are sent to core via a `set_preference` RPC call. Core persists the changes. This lets core handle file modification in a way that preserves document structure (https://github.com/joelself/tomllib), and is also opt-in for the client (or the user): if either wants to just edit files directly, that's no problem. It is also reasonably cross platform (a user can export their settings and have it just work), but is a bit cumbersome.

5. **Multimodal**: Allow the client to select from 1. and 2., above. Configuration is file based by default, but the client can opt into managing preferences, (with a launch flag or some RPC call) in which case core defers. This adds complexity in exchange for improved client flexibility.

## Conclusion

This is intended to sketch out some of my thoughts around persistence specifically. I would like to try and come up with some solution that accommodates various possible use cases, and I don't know what all of those are. If anything is missing or unclear, please help me clarify.

**Comment by jansol** [comment]
I think this comes a lot down to what xi-core aims to be. If it intends to be an (almost) standalone application that just needs a UI, then the traditional (1.) file-based approach makes a lot of sense. On the other hand if the goal is to be an embeddable text editor component usable in say, a UI framework, then the client-driven RPC-based approach (2.) sounds a lot more flexible.

**Comment by cmyr** [comment]
@jansol the goals I'm keeping in mind:
1. making it easy to write frontends
2. allowing frontends to conform to platform idioms
3. supporting a wide range of use cases, including but not limited to: heavily GUI editors/DEs, lightweight TUI editors, and being included as a component in a UI toolkit.
...
4. portability

These goals are at odds here, somewhat.  Ideally I'd like to find a solution that is a good balance between being easy to use, but also supporting use cases we may not be imagining.

**Comment by jansol** [comment]
Then I'd like to propose

6. **RPC-based, with a file-based shim included** i.e. (2.) but add an optional wrapper for it that implements (1.) by loading config files during init (or when requested) and does the necessary RPC calls.

This seems like a reasonable way to avoid the issues with shared config files forced by the core, while keeping it as a quick and dirty option for people who just want to get it up and running.

**Comment by cmyr** [comment]
@jansol That sounds like something closer to 4 or 5, depending on whether core does all the serialization. Are there any major differences I might be missing?

**Comment by jansol** [comment]
The idea is that the core does **only** RPC, to avoid any conflicts when embedding. For lazy client developers there is a wrapper lib that does the file<->RPC translation, that can be bundled with the core but is not strictly part of it.

**Comment by cmyr** [comment]
@jansol remember that user settings are optional; we work fine without them. This is important in the embedded case, where I don't think we would be exposing user settings at all; if customization is needed, that can be handled by per-session overrides (for which an RPC mechanism will be necessary anyway, and which don't persist between launches). 

If we think that still isn't enough, then we _do_ have the option of 5., which wasn't really explained very much; with 5, the client can opt in to manual management, in which case core doesn't do anything with files at all, and lets the client be the Seat of Truth for persistent user settings. 

If I'm understanding your shim idea, it would be that we provide a library which exposes an RPC interface, and writes files behind the scenes? Or do you mean a library which watches some config files and translates changes into RPC calls?

**Comment by sp3d** [comment]
@cmyr asked in irc for my thoughts, so after taking my damn time (sorry!) I'm glad to oblige:

tl;dr: the core should only communicate over RPC, and the frontend should manage persistence, because this (Approach 2) requires minimal complexity in the core, avoids having multiple core<->frontend comm channels, allows frontends to handle challenges of platform integration and exotic environments, and allows independent frontends to coexist sans conflict.


Here are my thoughts about the individual approaches proposed.

Approaches 1 (File based, core managed) and 4 (File backed, RPC modified) would need some way to inform core of a way to namespace settings, or else multiple consumers of Xi-core (such as a sublime-style editor and a GUI toolkit) would interfere with each other. For example, the client might pass a file or directory path into the core as a configuration namespace. The core still has to wait for the client before it can resolve configuration, but can save state on its own afterward. This seems moderately complicated, and file-watching is a racy, fiddly bit of I/O with capabilities that vary a lot by platform.

Approach 2 seems the simplest to me; it keeps the core from having to worry about filesystems and avoids interference between multiple consumers of the Xi core. Frontends can store settings in platform-native locations and inform the core of DE- or OS-provided settings, and only at most one party ever worries about files.

Approach 3 is as flexible as having the client directly send configuration values to the core, but routes needlessly through the filesystem, which opens up the possibility of any other mangling the files on the way, and would require some kind of synchronization barrier in order to know which RPCs will meet which state of on-disk configuration (although ultimately it's not possible to prevent some other process from modifying config between one party writing and the other reading). I don't think this is a good choice, even though it could plausibly work in "well-behaved" systems. Robustness comes from simplicity.

Approach 5 is reasonable, but I really don't think that "read a file and dump it into the RPC stream at startup" is a large enough burden on frontend authors to warrant offload and another set of codepaths in core.

**Comment by trishume** [comment]
I'd like to weigh in in favour of "File backed, RPC modified". Re @sp3d's comments:

- I think "minimal complexity in the core" is not a goal we want, at least when the tradeoff is complexity in the front end, everything that doesn't make things significantly worse if it is in the core should be, to avoid duplication and ease frontend dev.
- I don't understand what kind of scenario would have multiple consumers of xi-core interfering. Your parenthetical doesn't make sense to me. It's my understanding that only one front end would be connected to Xi at a time.

In favour of an approach involving files and the core:
- I really like being able to version my configs in Git, and comment my settings.
- Custom settings should be really easy for plugin authors to add, like in Sublime how you just read a key.

Also I'd like to point out that settings may also be related to the core or the frontend, or a plugin. Here's a few options to do this:
- It could be a separate dimension, each one involving all the listed levels. Alternatively, plugins could have fewer levels, if you say wanting to customize a plugin per syntax type is not needed.
- It could be one set of keys, like Emacs. Having one set of keys requires plugins and frontends to namespace themselves, but is simple.
- It could be based on names, like Sublime. In Sublime, syntax-specific settings are under the name of the syntax, plugin-specific settings under the name of the plugin, global settings under `Preferences`. Each has default, default for OS and user levels. Plugin settings resolve directly from the name. View settings resolve first from direct API/GUI changes for that view, then syntax-specific settings, then the global settings. 

I like how the Sublime approach balances simplicity and power. The amount of config levels you suggest scares me a bit, especially if combined with a dimension of plugin vs. frontend vs. core. I suggest trying to think of maybe a more general mechanism, although I can't think of one much better than the Sublime way right now.

**Comment by eyelash** [comment]
> I don't understand what kind of scenario would have multiple consumers of xi-core interfering. Your parenthetical doesn't make sense to me. It's my understanding that only one front end would be connected to Xi at a time.

@trishume at least on Linux we talked about building the front-end as a library that could then be used by every app that wants to display a text editor widget. There are may more use cases for a library providing a text editor widget than just a text editor. Things such as a git GUI embedding a diff view or a [GUI inspector embedding a CSS editor](https://blog.gtk.org/files/2017/04/css-view.png) come to my mind. In this scenario the possibility arises that you have multiple Xi clients (e.g. a text editor and a git GUI) running at the same time.

**Comment by trishume** [comment]
@eyelash Okay that makes sense, although I would think that would be implemented with each app having their own instance of `xi-core`, so apps don't interfere with each other. I also don't see how multiple apps embedding Xi would end up changing settings files in conflicting ways. In that use case, I think having my settings be the same across every app embedding Xi would be a feature.

Also I came up with a more general simpler mechanism for scoping preferences:

At the top level, there are multiple settings domains, these are entirely separate key spaces for settings, purely to avoid collisions. There's a domain for the core, a domain for each front end, and a domain for each plugin. These are like Sublime's settings names.

We use the mechanism of tagging buffers with scopes like I wrote about [in this RFC](https://github.com/google/xi-editor/blob/plugin-tag-rfc/doc/rfcs/plugin-lifetimes.md). Settings files would have sections and each section would have a buffer selector it applied to. The empty selector is global. 

Buffers are auto-tagged with their syntax so for syntax specific you can write selectors like `source.php` and even use levels like `source.html` and `source.html.erb`. Buffers maybe are also auto-tagged with platform and frontend so you can match on `os.mac` or `frontend.xi-win`.

Not sure how best to fit in user overrides here. Either make it a separate special thing, or just define a package load order that determines precedence, and the user's settings are always last. Or all buffers could be auto-tagged with `user` and adding `user` to your selectors would make them bind more strongly and thus override other settings for that selector, simplest but kinda hacky.

I like this solution since it's both more powerful than other proposed solutions, more general, and most of the work (buffer tagging, selector matching) would be shared with the plugin activation system, leading to less complexity overall.

**Comment by eyelash** [comment]
@trishume yes, each app would have its own instance of `xi-core` but what if two apps change the settings at the same time (assuming settings can be changed though the front-end). I guess it is unlikely and I guess it can be solved but the scenario still has to be considered.

**Comment by trishume** [comment]
@eyelash Yah it's pretty unlikely, since the user would have to perform an action in two apps in less than the time it takes to write a text file (milliseconds), and you can solve it by just using the atomic rename method of writing a file.

**Comment by cmyr** [comment]
@eyelash @trishume and if the user didn't desire to be sharing preferences, they would could configure a different`XI_CONFIG_DIR` in the frontend.

**Comment by jansol** [comment]
So if say, Qt uses xi to implement editable text fields and you make an application with multiple such fields and they need different settings (say, a code field and a comment field next to each other), how would that work? The app developer has to define a different `XI_CONFIG_DIR` for every text field?

**Comment by cmyr** [comment]
@jansol in that case you would probably use the RPC system to set per-session/view overrides as needed.

**Comment by jansol** [comment]
But if it was file-backed by the core, the fields would overwrite each other's configs all the time.

**Comment by trishume** [comment]
@jansol In @cmyr's original comment, that's what "Session overrides (system)" are for, those don't need files, those are done via an API and not persisted.

In my proposal, there would be an API for adding tags to buffers, the same API used for manipulating plugin activation, where you could tag buffers as a code field or comment field, then have a settings file that defines what the behaviours of each of those types of field are.

**Comment by cmyr** [comment]
@trishume I'm having a bit of a slow saturday. I agree that some mechanism for tagging probably makes sense.

>I like how the Sublime approach balances simplicity and power. The amount of config levels you suggest scares me a bit, especially if combined with a dimension of plugin vs. frontend vs. core. I suggest trying to think of maybe a more general mechanism, although I can't think of one much better than the Sublime way right now.

ST was my primary influence in terms of thinking how these various sources should stack. I think we have essentially the same number of config levels: general preferences (overridable), platform settings, syntax-specific settings (overridable) and per-view settings (overridable).

One major thing we haven't really discussed is how syntaxes in xi will be handled/described/added. 

Plugin preferences, in the sense of say setting options in a `rustfmt` plugin, exist outside of this hierarchy. We don't have a mechanism for plugins to add a new syntax description, right now. I'm sort of hoping to avoid thinking about that just yet, but maybe this is irresponsible of me 😕

**Comment by trishume** [comment]
@cmyr I have some thoughts in [my RFC](https://github.com/google/xi-editor/blob/plugin-tag-rfc/doc/rfcs/plugin-lifetimes.md), but basically it comes down to syntax highlighting plugins being an extension of plugin binding. Plugins like syntect that can load grammar files can use `walkdir` to search the config directory for all `.sublime-syntax` files and use those as well.

**Comment by cmyr** [comment]
@trishume I think a limitation of that approach is that it constrains us to syntect-like (regex based) syntax definitions. I think ultimately we desire a more expansive idea of what a syntax is, and how it can be described? Although again I really haven't given this a ton of thought.

One thing that in general we're going to have to think about is how tightly we wish to be coupled to syntect. For instance my current most wanted feature is auto-indentation. There's a reasonable case that plugins providing syntax highlighting should also be providing indentation, and in that case there's a good argument for adding indentation handling (and generally handling `.tmPreferences` files) to `syntect`.

**Comment by trishume** [comment]
@cmyr why does it constrain Xi to regex-based syntaxes? The method I described should work just fine for other highlighters. In those cases you just have a plugin that provides highlighting for some languages directly instead of loading grammar files.

**Comment by cmyr** [comment]
@trishume you're right, I was misremembering how sublime represented syntaxes.

**Comment by cmyr** [comment]
Closing this for now, we have an approximately functioning config system as of #433 or so.
